### PR TITLE
Support native Grouped-Query Attention (GQA) Key/Value head broadcast.

### PR DESCRIPTION
**Pull request by gaga1313**
Currently, when downstream models utilizing Grouped-Query Attention (GQA), such as Gemma 3, Gemma 2, or Gemma call `keras.ops.dot_product_attention` where num_query_heads > num_key_value_heads (and num_kv_heads > 1), Keras Core encounters a fatal runtime shape mismatch across both PyTorch and TensorFlow backends:

**PyTorch Backend:** _RuntimeError: The size of tensor a (8) must match the size of tensor b (4) at non-singleton dimension 1_
**TensorFlow Backend:** _InvalidArgumentError: Expected dimension 4 at axis 2 of the input but got dimension 2 [Op:Einsum]_

In `GQA`, multiple Query heads share a smaller number of Key/Value heads. When executing `dot_product_attention`, Keras Core delegates the mathematical blending to backend-specific routines. In the PyTorch backend (torch/nn.py), it hands the tensors to `torch.nn.functional.scaled_dot_product_attention` (SDPA). In the TensorFlow backend (tensorflow/nn.py), it hands them to an XLA-optimized tf.einsum("BTNH,BSNH->BNTS"). Both underlying operations rigidly expect the Query and Key tensors to have either equal head dimensions or be automatically broadcastable (which only works for Multi-Query Attention where num_kv_heads == 1). Because num_kv_heads > 1 in GQA configurations, both PyTorch and TensorFlow fail to broadcast and throw explicit shape mismatch crashes.

This PR updates GQA Key/Value head replication in the `dot_product_attention` for successful GQA execution on the affected backends. 

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

File: keras/src/backend/tensorflow/nn.py [hunk 1]
```diff
@@ -1754,6 +1754,19 @@ def dot_product_attention(
...
+
+    num_query_heads = query.shape[2]
+    num_kv_heads = key.shape[2]
+    if (
+        num_query_heads is not None
+        and num_kv_heads is not None
+        and num_query_heads > num_kv_heads
+        and num_kv_heads > 1
+    ):
+        groups = num_query_heads // num_kv_heads
```
Comment (hertschuh) [inline keras/src/backend/tensorflow/nn.py:1766]: What if `num_query_heads` is not divisible by `num_kv_heads`?

File: keras/src/backend/torch/nn.py [hunk 2]
```diff
@@ -1485,6 +1485,13 @@ def dot_product_attention(
     key = torch.transpose(key, axis0, axis1)
     value = torch.transpose(value, axis0, axis1)
 
+    num_query_heads = query.shape[1]
+    num_kv_heads = key.shape[1]
+    if num_query_heads > num_kv_heads and num_kv_heads > 1:
+        groups = num_query_heads // num_kv_heads
+        key = torch.repeat_interleave(key, repeats=groups, dim=1)
+        value = torch.repeat_interleave(value, repeats=groups, dim=1)
```
Comment (hertschuh) [inline keras/src/backend/torch/nn.py:1493]: Why `repeat_interleave` rather than just `repeat`?

Comment (gaga1313) [reply keras/src/backend/torch/nn.py:1493]: Oh I did this because in GQA adjacent group of Query heads share a single Key/Value head.

File: keras/src/backend/tensorflow/nn.py (hunk 1 above)
Comment (gaga1313) [reply keras/src/backend/tensorflow/nn.py:1766]: Ideally a user should set these such that num_query_heads is divisible by num_kv_heads

If it isn't divisible then the k/v heads will be shared non-uniformly among query heads. I am not sure we want to support that. 

Our current changes in torch and tensorflow, reach parity with JAX.
```
Jax Backend
# _dot_product_attention_xla
B, T, N, H = query.shape
G = N // K
query = jnp.reshape(query, (B, T, K, G, H))
```

The reshape will throw an error.


The original paper doesn't restrict towards having uniform distribution https://arxiv.org/pdf/2305.13245, but the idea is to distribute queries to few k/v heads to save KV cache space.

**Review by gaga1313 — COMMENTED** [review]
Answered comments.

**Comment by hertschuh** [comment]
We should be able to add the same feature to the other backends, right? (Numpy and OpenVino)
